crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -1753,6 +1758,30 @@ impl IndexerReader {
             .await
     }
 
+    pub async fn multi_get_transaction_block_response_in_blocking_task_with_preserved_order(
+        &self,
+        ordered_digests: Vec<TransactionDigest>,
+        options: iota_json_rpc_types::IotaTransactionBlockResponseOptions,
+    ) -> Result<Vec<IotaTransactionBlockResponse>, IndexerError> {
+        let order_map: HashMap<TransactionDigest, usize> = ordered_digests
+            .iter()
+            .enumerate()
+            .map(|(index, &id)| (id, index))
+            .collect();
+
+        Ok(self
+            .multi_get_transaction_block_response_in_blocking_task_impl(&ordered_digests, options)
+            .await?
+            .into_iter()
+            .sorted_by_key(|tx| {
+                order_map
+                    .get(&tx.digest)
+                    .copied()
+                    .expect("All digests should have some order")
+            })
+            .collect::<Vec<_>>())
```

Comment:
You can avoid consuming and recreating the vector, by sorting the result of the query in place. Something like

```rust
let mut transactions = self.multi_get().await?;
transactions.sort_unstable_by_key();
return Ok(transactions);
```

You can safely use `sort_unstable` as we sort by distinct keys.